File: frontend_v2/src/features/financial/lib/ledger-graph.ts
import { z } from "zod"
export const postingGraph = z
  .object({
    case_id: z.string(),
    account_id: z.string().nullable(),
    start_date: z.string().nullable(),
    end_date: z.string().nullable(),
    population: z.enum(["working", "verified"]),
    snapshot_sha256: z.string().regex(/^[a-f0-9]{64}$/),
    applied: z.literal(false),
    limitation: z.string(),
    excluded_rows: z.number().int().nonnegative(),
    nodes: z
      .array(
        z.object({
          id: z.string(),
          kind: z.enum(["account", "source_label"]),
          account_id: z.string(),
          label: z.string(),
        })
      )
      .max(2000),
    edges: z
      .array(
        z.object({
          id: z.string(),
          source: z.string(),
          target: z.string(),
          transaction_id: z.string(),
          source_document_id: z.string(),
          currency: z.string(),
          amount_minor: z.string().regex(/^(0|[1-9][0-9]*)$/),
          direction: z.enum(["credit", "debit"]),
          ordering_date: z.string(),
          description: z.string().nullable(),
          proof_class: z.enum(["p0", "p1", "p2", "p3"]),
        })
      )
      .max(1000),
  })
  .superRefine((data, ctx) => {
    const nodes = new Map(data.nodes.map((n) => [n.id, n]))
    const ids = new Set<string>()
    if (nodes.size !== data.nodes.length)
      ctx.addIssue({ code: "custom", message: "Repeated graph node." })
    for (const edge of data.edges) {
      const a = nodes.get(edge.source),
        b = nodes.get(edge.target)
      if (
        !a ||
        !b ||
        a.account_id !== b.account_id ||
        edge.id !== edge.transaction_id ||
        ids.has(edge.id) ||
        (edge.direction === "debit"
          ? a.kind !== "account" || b.kind !== "source_label"
          : a.kind !== "source_label" || b.kind !== "account")
      )
        ctx.addIssue({
          code: "custom",
          message: "Graph edge does not match its account and source group.",
        })
      ids.add(edge.id)
    }
  })
export type PostingGraph = z.infer<typeof postingGraph>

File: backend/services/financial/ledger_graph.py
"""Authoritative posting graph: source-label groups are not resolved identities."""
import hashlib
import json
from services.financial.ledger_summary import LedgerSummaryError
from services.financial.working_totals import working_totals_from_readings



def ledger_posting_graph(export, *, population='working'):
    if population not in ('working', 'verified'):
        raise LedgerSummaryError('Choose working or verified readings.')
    document = json.loads(export.snapshot.content)
    ledger = document['ledger']
    if not document.get('export_ready') or not ledger.get('history_captured'):
        raise LedgerSummaryError('A posting graph requires a consistent ledger and source history.')
    totals = working_totals_from_readings(ledger) if population == 'working' else ledger
    readings = [r for r in ledger['readings'] if (r['exclusion_reason'] in (None, 'proof_class_not_included') if population == 'working' else r['included'])]
    nodes, edges = {}, []
    for reading in readings:
        row = reading['row']; account = row.get('canonical_account_id') or row['account_id']; label = row.get('from_name' if row['direction'] == 'credit' else 'to_name', row['counterparty_raw'])
        if label is not None and not isinstance(label, str): raise LedgerSummaryError('A source counterparty label is invalid.')
        account_key = 'account:' + account
        account_details = reading.get('account', {})
        nodes.setdefault(account_key, dict(id=account_key, kind='account', account_id=account,
            label=row.get('canonical_account_label') or account_details.get('label') or ('Account ' + account[:8])))
        group_key = 'source-label:' + hashlib.sha256(json.dumps([account, row['currency'], label], ensure_ascii=False).encode()).hexdigest()
        if not row.get('counterparty_link'):
            nodes.setdefault(group_key, dict(id=group_key, kind='source_label', account_id=account,
                label=label or 'Name not identified'))
        identity = row.get('counterparty_link')
        if identity:
            group_key = identity['kind'] + ':' + identity['id']
            nodes.setdefault(group_key, dict(id=group_key, kind='account' if identity['kind'] == 'account' else 'source_label', account_id=identity['id'] if identity['kind'] == 'account' else account, label=identity['label']))
        source, target = (account_key, group_key) if row['direction']=='debit' else (group_key, account_key)
        edges.append(dict(id=row['key'], source=source, target=target, transaction_id=row['key'],
            source_document_id=row['source_document_id'], currency=row['currency'], amount_minor=row['amount_minor'],
            direction=row['direction'], ordering_date=row['ordering_date'], description=row['description'],
            proof_class=row['proof_class'], category=row.get('category', ''),
            label_sources=row.get('label_sources', {})))
    return dict(case_id=ledger['case_id'], account_id=ledger['account_id'], start_date=ledger['start_date'], end_date=ledger['end_date'],
        population=population, snapshot_sha256=export.snapshot.sha256, applied=False,
        nodes=sorted(nodes.values(), key=lambda n:n['id']), edges=edges, currencies=totals['currencies'], excluded_rows=totals['excluded_rows'],
        limitation='One edge per current posting. Arrows show the account posting direction. Names may come from the statement, investigator edits, or description suggestions; each edge records its label origins. Name groups are local to each account and currency; equal names do not establish identity. Missing labels do not identify a party. Transfers are not paired in this graph; use the separate transfer scenario for explicit pairing assumptions.')
